File: typescript/tests/reconcile.test.ts
import { describe, it, expect, vi } from "vitest";
import { decideFulfillment, reconcileTransaction } from "../src/lib/reconcile.js";
import { PagouHttpClient } from "../src/lib/http.js";
import type { PagouConfig } from "../src/lib/config.js";

const config: PagouConfig = {
  environment: "sandbox",
  baseUrl: "https://api.sandbox.pagou.ai",
  apiToken: "test",
  timeoutMs: 1000,
  maxRetries: 0,
};

function jsonResponse(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), { status, headers: { "content-type": "application/json" } });
}

describe("decideFulfillment", () => {
  it("fulfills only on settled statuses", () => {
    expect(decideFulfillment("paid")).toBe("fulfill");
    expect(decideFulfillment("captured")).toBe("fulfill");
  });

  it("waits on non-terminal statuses", () => {
    expect(decideFulfillment("pending")).toBe("wait");
    expect(decideFulfillment("three_ds_required")).toBe("wait");
    expect(decideFulfillment("processing")).toBe("wait");
  });

  it("cancels on terminal failures", () => {
    expect(decideFulfillment("expired")).toBe("cancel");
    expect(decideFulfillment("refused")).toBe("cancel");
    expect(decideFulfillment("canceled")).toBe("cancel");
  });
});

describe("reconcileTransaction", () => {
  it("fetches the transaction and returns a fulfillment decision", async () => {
    const fetchImpl = vi.fn(async () =>
      jsonResponse(200, { success: true, requestId: "r", data: { id: "tx_1", status: "paid" } }),
    );
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    const result = await reconcileTransaction("tx_1", client);
    expect(result?.decision).toBe("fulfill");
    expect(result?.transaction.status).toBe("paid");
  });

  it("returns null when the transaction does not exist", async () => {
    const fetchImpl = vi.fn(async () => jsonResponse(404, { message: "not found" }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    expect(await reconcileTransaction("missing", client)).toBeNull();
  });
});

File: go/tests/reconcile_test.go
package tests

import (
	"net/http"
	"testing"

	"github.com/PagouAi/examples/go/internal/pagou"
)

func TestDecideFulfillment(t *testing.T) {
	fulfill := []pagou.TransactionStatus{"paid", "captured"}
	for _, s := range fulfill {
		if got := pagou.DecideFulfillment(s); got != pagou.Fulfill {
			t.Errorf("DecideFulfillment(%s) = %s, want fulfill", s, got)
		}
	}
	wait := []pagou.TransactionStatus{"pending", "three_ds_required", "processing"}
	for _, s := range wait {
		if got := pagou.DecideFulfillment(s); got != pagou.Wait {
			t.Errorf("DecideFulfillment(%s) = %s, want wait", s, got)
		}
	}
	cancel := []pagou.TransactionStatus{"expired", "refused", "canceled"}
	for _, s := range cancel {
		if got := pagou.DecideFulfillment(s); got != pagou.Cancel {
			t.Errorf("DecideFulfillment(%s) = %s, want cancel", s, got)
		}
	}
}

func TestReconcileTransactionReturnsDecision(t *testing.T) {
	client := newClient(testConfig(), func(*http.Request) (*http.Response, error) {
		return jsonResponse(200, `{"success":true,"requestId":"r","data":{"id":"tx_1","status":"paid"}}`, nil), nil
	})

	result, err := pagou.ReconcileTransaction(client, "tx_1")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result == nil || result.Decision != pagou.Fulfill {
		t.Fatalf("decision = %v, want fulfill", result)
	}
	if result.Transaction.Status != "paid" {
		t.Errorf("status = %q, want paid", result.Transaction.Status)
	}
}

func TestReconcileTransactionReturnsNilWhenMissing(t *testing.T) {
	client := newClient(testConfig(), func(*http.Request) (*http.Response, error) {
		return jsonResponse(404, `{"message":"not found"}`, nil), nil
	})

	result, err := pagou.ReconcileTransaction(client, "missing")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != nil {
		t.Errorf("result = %v, want nil", result)
	}
}
